### PR TITLE
[INFRA/IIGO] Sanctions, Pardons and Historical retribution integration [4/4]

### PR DESCRIPTION
**Pull request by Neelesh99**
# Summary

As per comments on #188, we've split up the PR.
This PR brings all the components together and adds any remaining testing.

## Additional Information

Part 4 of 4
Pls merge last

## Test Plan

Unit tests added.

### CODE REVIEW

**Comment by QFSW** [comment]
# ⚠️Warning! This diff contains changes in protected folders

**Comment by lhl2617** [comment]
Are all these 4 PRs linearly dependent? If they are not, you can [drop irrelevant commits](https://til.hashrocket.com/posts/iw7diudcph-dropping-commits-with-git-rebase) in their respective branch. This means that each PR is cleanly with their own commit.

I have no idea which commits to review on this, for example, unless you make 2-4 drafts and merge them sequentially after review.

(In a perfect world we'd be really mindful of our commit messages and use the rebase-amend workflow so we can review commits separately in one PR).

**Comment by lhl2617** [comment]
i just read the PR for 1/4 --- ignore my previous comment

also, instead of merging from 1-3 (which adds an extra, fugly, commit, you can use `rebase`

say I have branches A & B
```
A: 1---2---3---4
        \
B:       5---6
```
```
git checkout B
git rebase A
```
will result in
```
A: 1---2---3---4
                \
B:               5---6
```

**Comment by Neelesh99** [comment]
> Are all these 4 PRs linearly dependent? If they are not, you can [drop irrelevant commits](https://til.hashrocket.com/posts/iw7diudcph-dropping-commits-with-git-rebase) in their respective branch. This means that each PR is cleanly with their own commit.
> 
> I have no idea which commits to review on this, for example, unless you make 2-4 drafts and merge them sequentially after review.
> 
> (In a perfect world we'd be really mindful of our commit messages and use the rebase-amend workflow so we can review commits separately in one PR).

Unfortunately, they are linearly dependent, I tried to keep each commit to the sections that were mentioned in the first PR, but that necessitated code that was in PR's before it.

**Comment by Neelesh99** [comment]
> i just read the PR for 1/4 --- ignore my previous comment
> 
> also, instead of merging from 1-3 (which adds an extra, fugly, commit, you can use `rebase`
> 
> say I have branches A & B
> 
> ```
> A: 1---2---3---4
>         \
> B:       5---6
> ```
> 
> ```
> git checkout B
> git rebase A
> ```
> 
> will result in
> 
> ```
> A: 1---2---3---4
>                 \
> B:               5---6
> ```
Sure, I'll use that in the future

File: .gitignore [hunk 1]
```diff
@@ -20,4 +20,6 @@ output
 
 .vscode
 
-*.wasm
\ No newline at end of file
+*.wasm
+
+*.idea
```
Comment (lhl2617) [inline .gitignore:25]: press F for jetbrains

File: internal/common/baseclient/iigo.go [hunk 2]
```diff
@@ -83,12 +83,17 @@ func (c *BaseClient) TaxTaken(shared.Resources) {
 	// Just an update. Ignore
 }
 
-// GetTaxContribution FIXME: Add documentation. What does this function do?
+// GetTaxContribution gives value of how much the island wants to pay in taxes
```
Comment (lhl2617) [inline internal/common/baseclient/iigo.go:50]: can we get a new PR that fixes #160

File: internal/common/rules/ruleutilities.go [hunk 3]
```diff
@@ -1,18 +1,20 @@
...
-	if _, ok := VariableMap[variableName]; ok {
+	if _, ok := variableMap[variableName]; ok {
 		for k, v := range ruleStore {
 			_, found := searchForVariableInArray(variableName, v.RequiredVariables)
 			if found {
 				Rules = append(Rules, k)
 			}
 		}
 		return Rules, true
+	} else {
```
Comment (lhl2617) [inline internal/common/rules/ruleutilities.go]: listen to linter

File: internal/server/iigo.go [hunk 4]
```diff
@@ -35,12 +35,14 @@ func (s *SOMASServer) runIIGOTax() error {
...
 		err := s.takeResources(clientID, tax, "tax")
-		if err == nil {
-			s.gameState.CommonPool += tax
+		err2 := s.takeResources(clientID, sanction, "sanction")
+		if err == nil && err2 == nil {
+			s.gameState.CommonPool += tax + sanction
 			s.clientMap[clientID].TaxTaken(tax)
 		} else {
-			s.logf("Error getting tax from %v: %v", clientID, err)
+			s.logf("Error getting tax or sanctions from %v: %v sanction error: %v", clientID, err, err2)
```
Comment (lhl2617) [inline internal/server/iigo.go]: I'd handle them one by one right after getting the errs

File: internal/server/iigointernal/judiciary.go [hunk 5]
```diff
@@ -81,11 +81,35 @@ func (j *judiciary) PayPresident() shared.Resources {
...
 }
 
-// InspectHistory checks all actions that happened in the last turn and audits them.
+// inspectHistory checks all actions that happened in the last turn and audits them.
 // This can be overridden by clients.
 func (j *judiciary) inspectHistory(iigoHistory []shared.Accountability) (map[shared.ClientID]roles.EvaluationReturn, bool) {
 	j.budget -= serviceCharge
-	return j.clientJudge.InspectHistory(iigoHistory)
+	success := false
+	tempResults := map[shared.ClientID]roles.EvaluationReturn{}
```
Comment (lhl2617) [inline internal/server/iigointernal/judiciary.go]:
```suggestion
	var tempResults map[shared.ClientID]roles.EvaluationReturn
```

File: internal/server/iigointernal/judiciary.go [hunk 6]
```diff
@@ -81,11 +81,35 @@ func (j *judiciary) PayPresident() shared.Resources {
...
+	finalResults := getBaseEvalResults()
+	if j.clientJudge.HistoricalRetributionEnabled() {
+		for _, v := range j.localHistoryCache {
+			res, rsuccess := j.clientJudge.InspectHistory(v)
+			if !rsuccess {
+				success = false
```
Comment (lhl2617) [inline internal/server/iigointernal/judiciary.go]: never used

File: internal/server/iigointernal/judiciary.go [hunk 7]
```diff
@@ -81,11 +81,35 @@ func (j *judiciary) PayPresident() shared.Resources {
...
+					curr := finalResults[key]
+					curr.Evaluations = append(curr.Evaluations, accounts.Evaluations...)
+					curr.Rules = append(curr.Rules, accounts.Rules...)
+					finalResults[key] = curr
+				}
+			}
+		}
+		j.localHistoryCache = defaultInitLocalHistoryCache(historyCacheDepth)
+	}
+	tempResults, success = j.clientJudge.InspectHistory(iigoHistory)
```
Comment (lhl2617) [inline internal/server/iigointernal/judiciary.go]: `success` replaced here

File: internal/server/iigointernal/sanctionevaluator.go [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+			return nil, 0.0, errors.Errorf("At auxiliary vector entry: '%v' aux value outside of 0-3: "+
+				"'%v' was found", i, interpret)
+		}
+		resultVect = append(resultVect, res)
+	}
+
+	return resultVect, outputVal, nil
+}
+
+func checkAllValid(arr []bool) bool {
```
Comment (lhl2617) [inline internal/server/iigointernal/sanctionevaluator.go:87]: a lot of unused stuff -- go through again?

Comment (Neelesh99) [reply internal/server/iigointernal/sanctionevaluator.go:87]: I've just gone through
All of this stuff is used. 
checkAllValid is called by evaluateSanction (line 19 same file) and evaluate sanction itself is called in judiciary line 205. Not sure why the linter is saying they aren't used?

File: internal/server/iigointernal/judiciary.go (hunk 6 above)
Comment (Neelesh99) [reply internal/server/iigointernal/judiciary.go]: Oh crap yeah, not sure what I was thinking. thanks for the catch, patched in latest commit.

File: internal/server/iigointernal/judiciary.go (hunk 7 above)
Comment (Neelesh99) [reply internal/server/iigointernal/judiciary.go]: As above

File: internal/server/iigointernal/judiciary.go (hunk 5 above)
Comment (Neelesh99) [reply internal/server/iigointernal/judiciary.go]: Thanks, changed

File: internal/server/iigointernal/sanctionevaluator.go (hunk 8 above)
Comment (lhl2617) [reply internal/server/iigointernal/sanctionevaluator.go:87]: welp, thats strange

File: internal/server/iigo.go (hunk 4 above)
Comment (Neelesh99) [reply internal/server/iigo.go]: Yeah you're right, changed in latest commit

File: internal/common/rules/ruleutilities.go (hunk 3 above)
Comment (Neelesh99) [reply internal/common/rules/ruleutilities.go]: Yeah fair enough, fixed now

File: .gitignore (hunk 1 above)
Comment (Neelesh99) [reply .gitignore:25]: Yeah turns out I'm very bad at rebasing and somehow it kept trying to pickup the idea files, so I had to make it stop somehow ;)

File: internal/common/baseclient/iigo.go (hunk 2 above)
Comment (Neelesh99) [reply internal/common/baseclient/iigo.go:50]: Ok yeah, thanks for bringing it to my attention. If I understand correctly the BASE|OPTIONAL|COMPULSORY note is with regards to whether it needs to be overridden?

Comment (lhl2617) [reply internal/common/baseclient/iigo.go:50]: yep :)

File: internal/common/rules/ruleutilities.go [hunk 9]
```diff
@@ -1,9 +1,9 @@
 package rules
 
 // PickUpRulesByVariable returns a list of rule_id's which are affected by certain variables.
-func PickUpRulesByVariable(variableName VariableFieldName, ruleStore map[string]RuleMatrix) ([]string, bool) {
+func PickUpRulesByVariable(variableName VariableFieldName, ruleStore map[string]RuleMatrix, variableMap map[VariableFieldName]VariableValuePair) ([]string, bool) {
 	var Rules []string
-	if _, ok := VariableMap[variableName]; ok {
+	if _, ok := variableMap[variableName]; ok {
 		for k, v := range ruleStore {
 			_, found := searchForVariableInArray(variableName, v.RequiredVariables)
```
Comment (lhl2617) [inline internal/common/rules/ruleutilities.go:8]: Did we not refactor some searches to return -1 when not found? 
@E-Mac27

Comment (lhl2617) [reply internal/common/rules/ruleutilities.go:8]: anyway, not within the scope of this diff :)

File: internal/server/iigo.go [hunk 10]
```diff
@@ -35,23 +35,46 @@ func (s *SOMASServer) runIIGOTax() error {
 	clientMap := getNonDeadClients(s.gameState.ClientInfos, s.clientMap)
 	for clientID, v := range clientMap {
 		tax := v.GetTaxContribution()
-		err := s.takeResources(clientID, tax, "tax")
-		if err == nil {
+		sanction := v.GetSanctionPayment()
+		clientTaxErr := s.takeResources(clientID, tax, "tax")
```
Comment (lhl2617) [inline internal/server/iigo.go:41]: rewrite to handle `err != nil` branch first

File: internal/server/iigo.go [hunk 11]
```diff
@@ -35,23 +35,46 @@ func (s *SOMASServer) runIIGOTax() error {
...
 			s.gameState.CommonPool += tax
 			s.clientMap[clientID].TaxTaken(tax)
+			taxPaid = tax
 		} else {
-			s.logf("Error getting tax from %v: %v", clientID, err)
+			s.logf("Error getting tax from %v: %v", clientID, clientTaxErr)
+			taxPaid = 0
+		}
+		clientSanctionErr := s.takeResources(clientID, sanction, "sanction")
+		if clientSanctionErr == nil {
```
Comment (lhl2617) [inline internal/server/iigo.go]: ditto, `err != nil` first.

File: internal/server/iigointernal/judiciary.go (hunk 7 above)
Comment (lhl2617) [reply internal/server/iigointernal/judiciary.go]: why not remove L88-89 and use `:=`?

File: internal/server/iigointernal/judiciary.go [hunk 12]
```diff
@@ -118,7 +140,96 @@ func (j *judiciary) appointNextPresident(currentPresident shared.ClientID, allIs
...
+	islandTransgressions := map[shared.ClientID][]string{}
+	for k, v := range j.evaluationResults {
+		islandTransgressions[k] = unpackSingleIslandTransgressions(v)
+	}
+	j.scoreIslandTransgressions(islandTransgressions)
+}
+
+// scoreIslandTransgressions uses the client provided ruleViolationSeverity map to score island transgressions
+func (j *judiciary) scoreIslandTransgressions(transgressions map[shared.ClientID][]string) {
+	for islandId, rulesBroken := range transgressions {
```
Comment (lhl2617) [inline internal/server/iigointernal/judiciary.go]: linter

File: internal/server/iigointernal/judiciary.go [hunk 13]
```diff
@@ -118,7 +140,96 @@ func (j *judiciary) appointNextPresident(currentPresident shared.ClientID, allIs
...
+	for islandID, sanctionScore := range j.sanctionRecord {
+		islandSanctionTier := getIslandSanctionTier(sanctionScore, j.sanctionThresholds)
+		sanctionEntry := roles.Sanction{
+			ClientID:     islandID,
+			SanctionTier: islandSanctionTier,
+			TurnsLeft:    sanctionLength,
+		}
+		currentSanctions = append(currentSanctions, sanctionEntry)
+	}
+	j.localSanctionCache[0] = currentSanctions
```
Comment (lhl2617) [inline internal/server/iigointernal/judiciary.go:180]: what if out of range

File: internal/server/iigointernal/judiciary.go [hunk 14]
```diff
@@ -118,7 +140,96 @@ func (j *judiciary) appointNextPresident(currentPresident shared.ClientID, allIs
...
+}
+
+// sanctionEvaluate allows the clients to effectively pardon islands, levy and communicate sanctions
+func (j *judiciary) sanctionEvaluate(reportedIslandResources map[shared.ClientID]shared.Resources) {
```
Comment (lhl2617) [inline internal/server/iigointernal/judiciary.go:184]: can't follow what's written here. break down into smaller fns.

File: internal/server/iigointernal/judiciary.go [hunk 15]
```diff
@@ -149,6 +260,15 @@ func (j *judiciary) clearHistoryCache() {
 
 // Helper functions //
 
+func decrementSanctionTime(sanctions map[int][]roles.Sanction) (updatedSanctions map[int][]roles.Sanction) {
+	for _, v := range sanctions {
+		for _, sanction := range v {
+			sanction.TurnsLeft -= 1
```
Comment (lhl2617) [inline internal/server/iigointernal/judiciary.go]: `--`? as per linter

File: internal/server/iigointernal/judiciary.go [hunk 16]
```diff
@@ -149,6 +260,15 @@ func (j *judiciary) clearHistoryCache() {
 
 // Helper functions //
 
+func decrementSanctionTime(sanctions map[int][]roles.Sanction) (updatedSanctions map[int][]roles.Sanction) {
```
Comment (lhl2617) [inline internal/server/iigointernal/judiciary.go:270]: can't find a test

File: internal/server/iigointernal/judiciary.go [hunk 17]
```diff
@@ -323,46 +454,33 @@ func checkSizes(sanctionCache map[int][]roles.Sanction, pardons map[int][]bool)
...
-	}
-}
-
-func searchForVariableInArray(val rules.VariableFieldName, array []rules.VariableFieldName) (int, bool) {
-	for i, v := range array {
-		if v == val {
-			return i, true
-		}
+func getBaseEvalResults() map[shared.ClientID]roles.EvaluationReturn {
+	return map[shared.ClientID]roles.EvaluationReturn{
```
Comment (lhl2617) [inline internal/server/iigointernal/judiciary.go]: use a for loop to populate

File: internal/server/iigointernal/judiciary.go [hunk 18]
```diff
@@ -193,6 +313,17 @@ func checkMonotonicityOfSanctionThresholds(clientSanctionMap map[roles.IIGOSanct
 	return true
 }
 
+// unpackSingleIslandTransgressions fetches the rule names of any broken rules as per EvaluationReturns
+func unpackSingleIslandTransgressions(evaluationReturn roles.EvaluationReturn) []string {
```
Comment (lhl2617) [inline internal/server/iigointernal/judiciary.go:324]: no test?

File: internal/server/iigointernal/judiciary.go (hunk 14 above)
Comment (lhl2617) [reply internal/server/iigointernal/judiciary.go:184]: + test

File: internal/server/iigointernal/judiciary.go [hunk 19]
```diff
@@ -410,3 +528,23 @@ func searchEvalReturnForRuleName(name string, evaluationReturn roles.EvaluationR
...
 }
+
+// mergeEvalResults takes two evaluation results and merges them to provide a unified set
+func mergeEvalResults(set1 map[shared.ClientID]roles.EvaluationReturn, set2 map[shared.ClientID]roles.EvaluationReturn) map[shared.ClientID]roles.EvaluationReturn {
+	for key, val := range set1 {
+		if set2Val, ok := set2[key]; ok {
+			defRules := set2Val.Rules
+			defEvals := set2Val.Evaluations
+			finalRules := append(defRules, val.Rules...)
+			finalEvals := append(defEvals, val.Evaluations...)
```
Comment (lhl2617) [inline internal/server/iigointernal/judiciary.go]: this is definitely going to bite you--if the `defRules` container has enough space, `defRules` itself is mutated.

https://tour.golang.org/moretypes/15

File: internal/server/iigointernal/judiciary.go [hunk 20]
```diff
@@ -410,3 +528,23 @@ func searchEvalReturnForRuleName(name string, evaluationReturn roles.EvaluationR
 	}
 	return false
 }
+
+// mergeEvalResults takes two evaluation results and merges them to provide a unified set
+func mergeEvalResults(set1 map[shared.ClientID]roles.EvaluationReturn, set2 map[shared.ClientID]roles.EvaluationReturn) map[shared.ClientID]roles.EvaluationReturn {
+	for key, val := range set1 {
+		if set2Val, ok := set2[key]; ok {
+			defRules := set2Val.Rules
```
Comment (lhl2617) [inline internal/server/iigointernal/judiciary.go:525]: why not write a `mergeEvaluationReturn` function?

File: internal/server/iigointernal/judiciary_test.go [hunk 21]
```diff
@@ -267,7 +396,7 @@ func TestPickUpRulesByVariable(t *testing.T) {
 // TestCheckPardons checks whether checkPardons is able to correctly identify pardons issued on sanctions that
 // don't exist, and then subtracts these pardoned sanctions from the sanctionCache
 func TestImplementPardons(t *testing.T) {
-	TeamIDs := []shared.ClientID{shared.Team1, shared.Team2, shared.Team3, shared.Team4, shared.Team5, shared.Team6}
+	TeamIDs := [...]shared.ClientID{shared.Team1, shared.Team2, shared.Team3, shared.Team4, shared.Team5, shared.Team6}
```
Comment (lhl2617) [inline internal/server/iigointernal/judiciary_test.go]: why is this here? `shared.TeamIDs` is exactly this.

**Review by lhl2617 — CHANGES_REQUESTED** [review]
note that small LoC is not the only metric for a good PR--tiny cognitive complexity is very favourable.

File: internal/server/iigointernal/judiciary.go (hunk 7 above)
Comment (Neelesh99) [reply internal/server/iigointernal/judiciary.go]: Oh yeah, sorry, fragments of code gone past

File: internal/server/iigo.go (hunk 10 above)
Comment (Neelesh99) [reply internal/server/iigo.go:41]: Sure, have fixed

File: internal/server/iigo.go (hunk 11 above)
Comment (Neelesh99) [reply internal/server/iigo.go]: THanks, yup, fixed now

File: internal/server/iigointernal/judiciary.go (hunk 13 above)
Comment (Neelesh99) [reply internal/server/iigointernal/judiciary.go:180]: The first line of the applySanctions function: "j.cycleSanctionCache()" (line 170), rebuilds the sanction cache and puts in an empty entry into j.localSanctionCache[0], this should mean that it isn't out of range.

File: internal/server/iigointernal/judiciary.go (hunk 18 above)
Comment (Neelesh99) [reply internal/server/iigointernal/judiciary.go:324]: Please see judiciary_test.go lines 622 to 672

File: internal/server/iigointernal/judiciary.go (hunk 13 above)
Comment (lhl2617) [reply internal/server/iigointernal/judiciary.go:180]: ok, i trust you to remember this one day before the deadline...

File: internal/server/iigointernal/judiciary.go (hunk 18 above)
Comment (lhl2617) [reply internal/server/iigointernal/judiciary.go:324]: ah -- missed it- sorry and thanks

File: internal/server/iigointernal/judiciary.go (hunk 20 above)
Comment (Neelesh99) [reply internal/server/iigointernal/judiciary.go:525]: Yeah that function is poorly named. Renaming to mergeEvaluationReturn, since that's essentially what it does.

File: internal/server/iigointernal/judiciary_test.go (hunk 21 above)
Comment (Neelesh99) [reply internal/server/iigointernal/judiciary_test.go]: Yeah sorry, relic of some old code I didn't update when I moved it to use shared.TeamIds in the judiciary.go function. Removed and refactored now.

File: internal/server/iigointernal/judiciary.go (hunk 19 above)
Comment (Neelesh99) [reply internal/server/iigointernal/judiciary.go]: Thanks for the catch, would never have known. Patched in latest commit

File: internal/server/iigointernal/judiciary.go (hunk 17 above)
Comment (Neelesh99) [reply internal/server/iigointernal/judiciary.go]: Yeah thanks, way easier.

File: internal/common/rules/ruleutilities.go (hunk 9 above)
Comment (E-Mac27) [reply internal/common/rules/ruleutilities.go:8]: Oh yeh I wasn't smart and I didn't return -1 and just ditched the bool. 

We technically don't need the bool but meh we left it there cause idk what was required from that search function. From what I seen, Index aren't even used.

File: internal/server/iigointernal/judiciary.go (hunk 18 above)
Comment (Neelesh99) [reply internal/server/iigointernal/judiciary.go:324]: Due to test added as above, this had been moved (lines 706 - 756)

File: internal/server/iigointernal/judiciary.go (hunk 16 above)
Comment (Neelesh99) [reply internal/server/iigointernal/judiciary.go:270]: Thanks for the catch, test added

File: internal/server/iigointernal/judiciary.go (hunk 14 above)
Comment (Neelesh99) [reply internal/server/iigointernal/judiciary.go:184]: Fair enough, I've broken into smaller functions. In terms of testing, since this relies on the globalRulesCache and other elements such as broadcasting communications. Is it ok if we hold the testing until integration tests for IIGO as a whole.

File: internal/server/iigointernal/judiciary.go (hunk 12 above)
Comment (Neelesh99) [reply internal/server/iigointernal/judiciary.go]: Thanks, changed

File: internal/server/iigointernal/judiciary.go (hunk 13 above)
Comment (Neelesh99) [reply internal/server/iigointernal/judiciary.go:180]: Haha, on my head be it. Tbf the localSanctionCache is an instantiated map anyway so at worst this is just assigning a value to a new key (even if it shouldn't be new)?

Comment (lhl2617) [reply internal/server/iigointernal/judiciary.go:180]: ah ok - it's a map. what does the key represent and why is it 0 here?

File: internal/server/iigointernal/judiciary.go (hunk 14 above)
Comment (lhl2617) [reply internal/server/iigointernal/judiciary.go:184]: if you guarantee it works :)

File: internal/server/iigointernal/judiciary.go (hunk 13 above)
Comment (Neelesh99) [reply internal/server/iigointernal/judiciary.go:180]: The key represents how many timesteps ago the sanction was levied. The 0 was since the sanctions assigned here are ones that came from rules that were evaluated in the current session of IIGO.

File: internal/server/iigointernal/executive.go [hunk 22]
```diff
@@ -162,12 +162,12 @@ func (e *executive) reset(val string) {
 //requestRuleProposal asks each island alive for its rule proposal.
 func (e *executive) requestRuleProposal() {
 	e.budget -= serviceCharge
-	var rules []string
+	var ruleProposals []string
 	for _, island := range getIslandAlive() {
-		rules = append(rules, iigoClients[shared.ClientID(int(island))].RuleProposal())
+		ruleProposals = append(ruleProposals, iigoClients[shared.ClientID(int(island))].RuleProposal())
```
Comment (NGovani) [inline internal/server/iigointernal/executive.go]: what happens if the proposed rule isn't a valid one. Is this checked later on?

File: internal/server/iigointernal/judiciary.go (hunk 14 above)
Comment (Neelesh99) [reply internal/server/iigointernal/judiciary.go:184]: I will guarantee that it works in the context of this PR, and will make sure to thoroughly test in the post-MVP orchestration where it will be called.

File: internal/server/iigointernal/judiciary.go [hunk 23]
```diff
@@ -81,11 +81,31 @@ func (j *judiciary) PayPresident() shared.Resources {
...
-// InspectHistory checks all actions that happened in the last turn and audits them.
+// inspectHistory checks all actions that happened in the last turn and audits them.
 // This can be overridden by clients.
 func (j *judiciary) inspectHistory(iigoHistory []shared.Accountability) (map[shared.ClientID]roles.EvaluationReturn, bool) {
 	j.budget -= serviceCharge
-	return j.clientJudge.InspectHistory(iigoHistory)
+	finalResults := getBaseEvalResults(shared.TeamIDs)
+	if j.clientJudge.HistoricalRetributionEnabled() {
+		for _, v := range j.localHistoryCache {
+			res, rsuccess := j.clientJudge.InspectHistory(v)
```
Comment (NGovani) [inline internal/server/iigointernal/judiciary.go]: does the struct passed into inspectHistory have a turn date, it might be beneficial for clients to know which turn history they are inpecting

File: internal/server/iigointernal/judiciary.go [hunk 24]
```diff
@@ -118,7 +138,71 @@ func (j *judiciary) appointNextPresident(currentPresident shared.ClientID, allIs
...
+	}
+	j.scoreIslandTransgressions(islandTransgressions)
+}
+
+// scoreIslandTransgressions uses the client provided ruleViolationSeverity map to score island transgressions
+func (j *judiciary) scoreIslandTransgressions(transgressions map[shared.ClientID][]string) {
+	for islandID, rulesBroken := range transgressions {
+		totalIslandTurnScore := roles.IIGOSanctionScore(0)
+		for _, ruleBroken := range rulesBroken {
+			if score, ok := j.ruleViolationSeverity[ruleBroken]; ok {
```
Comment (NGovani) [inline internal/server/iigointernal/judiciary.go:157]: is the ruleViolationSeverity a config parameter, i.e. do we decide or do the agents

File: internal/server/iigointernal/judiciary.go (hunk 13 above)
Comment (NGovani) [reply internal/server/iigointernal/judiciary.go:180]: I believe it represents turns since the sanctions have come into play, cycleSanctionCache moves this one up until the max turns a sanction can be a in play at which point it is removed

File: internal/server/iigointernal/judiciary.go [hunk 25]
```diff
@@ -149,6 +233,49 @@ func (j *judiciary) clearHistoryCache() {
 
 // Helper functions //
 
+func runEvaluationRulesOnSanctions(localSanctionCache map[int][]roles.Sanction, reportedIslandResources map[shared.ClientID]shared.Resources) map[shared.ClientID]shared.Resources {
```
Comment (NGovani) [inline internal/server/iigointernal/judiciary.go]: since this is going to be tested later as part of the integration, for now at least could get a description of what the function does a comment. Just so we know its intended purpose. There's alot of moving parts which are hard to keep a track of

File: internal/server/iigointernal/judiciary.go [hunk 26]
```diff
@@ -253,7 +391,7 @@ func defaultInitLocalHistoryCache(depth int) map[int][]shared.Accountability {
 	return returnMap
 }
 
-func generateEmptyCommunicationsMap(allTeamIds []shared.ClientID) map[shared.ClientID][]map[shared.CommunicationFieldName]shared.CommunicationContent {
+func generateEmptyCommunicationsMap(allTeamIds [6]shared.ClientID) map[shared.ClientID][]map[shared.CommunicationFieldName]shared.CommunicationContent {
```
Comment (NGovani) [inline internal/server/iigointernal/judiciary.go]: Having the 6 hard coded is to do with the number of islands correct. Should this value not be taken from the gamestate or config so that we can run games with a varying number of islands

File: internal/server/iigointernal/sanctionevaluator.go [hunk 27]
```diff
@@ -0,0 +1,93 @@
+package iigointernal
+
+import (
+	"github.com/SOMAS2020/SOMAS2020/internal/common/rules"
+	"github.com/SOMAS2020/SOMAS2020/internal/common/shared"
+	"github.com/pkg/errors"
+	"gonum.org/v1/gonum/mat"
+)
+
+func evaluateSanction(sanction rules.RuleMatrix, localVariableCache map[rules.VariableFieldName]rules.VariableValuePair) shared.Resources {
```
Comment (NGovani) [inline internal/server/iigointernal/sanctionevaluator.go:10]: while the rest of the functions in this file look simple enough and likely don't need it, this function seems complicated enough to justify at least one testcase. Apologies if you have already made a test case for it, i can't seem to find it

 Plus a comment on how it works would be appreciated

**Review by NGovani — CHANGES_REQUESTED** [review]

File: internal/server/iigointernal/executive.go (hunk 22 above)
Comment (Neelesh99) [reply internal/server/iigointernal/executive.go]: Not for now, considering the changes that will likely come with the invented rules a system as described would need to be implemented, but until it's in I'm unsure what value we'd get out of it.

Comment (Neelesh99) [reply internal/server/iigointernal/executive.go]: Screw it I added one anyway

File: internal/server/iigointernal/judiciary.go (hunk 23 above)
Comment (Neelesh99) [reply internal/server/iigointernal/judiciary.go]: Good idea, added to InspectHistory function signature.

File: internal/server/iigointernal/judiciary.go (hunk 24 above)
Comment (Neelesh99) [reply internal/server/iigointernal/judiciary.go:157]: Agents decide we have a Judge method called GetRuleViolationSeverity() which is used to load this (see judiciary.go line 38)

File: internal/server/iigointernal/judiciary.go (hunk 25 above)
Comment (Neelesh99) [reply internal/server/iigointernal/judiciary.go]: Yeah sorry, thanks for the spot. Added a description.

File: internal/server/iigo.go (hunk 10 above)
Comment (lhl2617) [reply internal/server/iigo.go:41]: Should be directly in the line below

File: internal/server/iigointernal/judiciary.go (hunk 26 above)
Comment (Neelesh99) [reply internal/server/iigointernal/judiciary.go]: Yeah good catch, I've set it to feed off the length of the shared.TeamIds, I though it would be the best length for the array initialiser.

File: internal/server/iigointernal/sanctionevaluator.go (hunk 27 above)
Comment (Neelesh99) [reply internal/server/iigointernal/sanctionevaluator.go:10]: Sure, added test, I'll add some testing for the rest of judiciary while I'm at it, might as well get that done now

**Comment by Neelesh99** [comment]
Sorry this PR has gained a lot of LOC again, but all of them are new tests I promise, hopefully testing as much as is testable without integration tests.

File: internal/server/iigointernal/judiciary.go (hunk 14 above)
Comment (Neelesh99) [reply internal/server/iigointernal/judiciary.go:184]: Ok it's tested now ;)

File: internal/common/baseclient/iigo.go (hunk 2 above)
Comment (Neelesh99) [reply internal/common/baseclient/iigo.go:50]: I think @preetl has addressed most of these in #217, I've pulled main into this branch now

File: internal/server/iigo.go (hunk 10 above)
Comment (Neelesh99) [reply internal/server/iigo.go:41]: Sure, changes pushed

File: internal/server/iigointernal/judiciary.go (hunk 25 above)
Comment (Neelesh99) [reply internal/server/iigointernal/judiciary.go]: I realised with a little adjustment we can add a test to this and move the RulesInPlay dependency upstream. The function is now tested.

**Review by NGovani — APPROVED** [review]
Alright I'm happy with pushing this to main if you think it will work. @lhl2617 if you could give it one more lookover that would great

File: internal/common/baseclient/iigo.go (hunk 2 above)
Comment (lhl2617) [reply internal/common/baseclient/iigo.go:50]: hmm, doesn't show for me here but it's fine--we can add later

File: internal/common/rules/globalruleregistration.go [hunk 28]
```diff
@@ -36,86 +36,173 @@ func registerDemoRule() {
 	// Check internal/clients/team3/client.go for an implementation of a basic evaluator for this rule
 }
 
+// RawRuleSpecification allows a user to use the CompileRuleCase function to build a rule matrix
+type RawRuleSpecification struct {
+	Name    string
+	ReqVar  []VariableFieldName
+	V       []float64
```
Comment (lhl2617) [inline internal/common/rules/globalruleregistration.go]: why not Values

File: internal/server/iigointernal/judiciary.go (hunk 20 above)
Comment (lhl2617) [reply internal/server/iigointernal/judiciary.go:525]: Ahh,I meant a function that just deals with two `roles.EvaluationReturn`. But if you tested this and it's fine, I'm OK with this.

File: internal/server/iigointernal/orchestration.go [hunk 29]
```diff
@@ -48,6 +48,9 @@ var TaxAmountMapExport map[shared.ClientID]shared.Resources
 // AllocationAmountMapExport is a local allocation map for checking of rules
 var AllocationAmountMapExport map[shared.ClientID]shared.Resources
 
+// SanctionAmountMapExport
+var SanctionAmountMapExport map[shared.ClientID]shared.Resources
```
Comment (lhl2617) [inline internal/server/iigointernal/orchestration.go:52]: hah, linter annoying

**Review by lhl2617 — APPROVED** [review]
approving to unblock, trusting @NGovani . Some nitpicks left

File: internal/common/rules/globalruleregistration.go (hunk 28 above)
Comment (Neelesh99) [reply internal/common/rules/globalruleregistration.go]: Agreed and changed

File: internal/server/iigointernal/orchestration.go (hunk 29 above)
Comment (Neelesh99) [reply internal/server/iigointernal/orchestration.go:52]: Agreed and changed
